File: src/base64/Base64Encoder.py
import base64


"""
    base64 characters
    [A-Z]   26
    [a-z]   26
    [0-9]   10
    +/      2
    Total:  64 = 2 ^ 6 Bit
"""

base64_chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"


def encode_base64(data: str) -> str:
    binary_data = ''.join(format(ord(char), '08b') for char in data)

    # pad the binary data to a multiple of 6
    while len(binary_data) % 6 != 0:
        binary_data += '0'

    # group the binary data into chunks of 6 bits and convert to decimal
    chunks = [int(binary_data[i:i+6], 2) for i in range(0, len(binary_data), 6)]

    # map decimal values to Base64 characters
    encoded_data = ''.join(base64_chars[i] for i in chunks)

    # add padding '=' characters if needed
    padding = (4 - len(encoded_data) % 4) % 4
    encoded_data += '=' * padding
    return encoded_data

# --------------------------------------------------


def decode_base64(data: str) -> str:
    pass


def builtin_encode(data: str) -> str:
    encoded_data = base64.b64encode(data.encode("utf-8")).decode("utf-8")
    return encoded_data


def builtin_decode(data: str) -> str:
    decoded_data = base64.b64decode(data.encode("utf-8")).decode("utf-8")
    return decoded_data


# if __name__ == "__main__":
input_data = "Hello World"
print(ord('A'))
print(format(ord('A'), '08b'))
encoded_data_1 = builtin_encode(input_data)
encoded_data_2 = builtin_encode(input_data)
print(encoded_data_1)
print(encoded_data_2)

File: src/base64/java/Base64Encoder.java
import java.util.Base64;

/**
 * [A-Z] 26
 * [a-z] 26
 * [0-9] 10
 * +/ 2
 * Total 64 = 2 ^ 6 bit
 */

public class Base64Encoder {
    private String base64Chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    public String encode(String data) {
        StringBuilder encodedData = new StringBuilder();
        StringBuilder binaryData = new StringBuilder();
        for (char c : data.toCharArray()) {
            binaryData.append(String.format("%8s", Integer.toBinaryString(c)).replace(' ', '0'));
        }

        // pad binary data
        while (binaryData.length() % 6 != 0) {
            binaryData.append('0');
        }

        // encode bytes
        for (int i = 0; i < binaryData.length(); i += 6) {
            String chunk = binaryData.substring(i, i + 6);
            int decimalVal = Integer.parseInt(chunk, 2);
            encodedData.append(base64Chars.charAt(decimalVal));
        }

        // add padding '=' if needed
        int padding = (4 - encodedData.length() % 4) % 4;
        for (int i = 0; i < padding; i++) {
            encodedData.append('=');
        }

        return encodedData.toString();
    }

    public byte[] decode(String encodedData) {
        byte[] decodedData = null;

        return decodedData;
    }

    public static void main(String args[]) {
        System.out.println("Program started");

        Base64Encoder encoder = new Base64Encoder();
        String input = "Hello World!";
        String myEncoded = encoder.encode(input);
        byte[] myDecoded = encoder.decode(myEncoded);

        String encoded = Base64.getEncoder().encodeToString(input.getBytes());
        System.out.println(String.format("Base64 Encoder : %s", myEncoded));
        System.out.println(String.format("Base64 Library : %s", encoded));
    }
}